src/implicit.rs

```diff
@@ -255,30 +310,40 @@ pub fn resolve_neutral<D: BidiDataSource>(
         if let Some(class_to_set) = class_to_set {
             // update all processing classes corresponding to the start and end elements, as requested.
             // We should include all bytes of the character, not the first one.
-            let end_len_utf8 = text[pair.start..].chars().next().unwrap().len_utf8();
+            let end_len_utf8 = text[pair.end..].chars().next().unwrap().len_utf8();
             for class in &mut processing_classes[pair.start..pair.start + start_len_utf8] {
                 *class = class_to_set;
             }
             for class in &mut processing_classes[pair.end..pair.end + end_len_utf8] {
                 *class = class_to_set;
             }
+            // <https://www.unicode.org/reports/tr9/#Retaining_Explicit_Formatting_Characters>
+            for idx in sequence.iter_backwards_from(pair.start, pair.start_run) {
+                let class = &mut processing_classes[idx];
+                if *class != BN {
+                    break;
+                }
+                *class = class_to_set;
```

Comment:
You're only setting one element; is this intentional, or do you need to set multiple elements? Is `processing_classes` in UTF-8 or UTF-32 indices?

Reply:
The loop makes this unnecessary; we are maintaining the property of every byte having its classes set so we don't have to worry about this. The existing code does this in many spots. I do want to clean this up by aggressively commenting or by building better abstractions. (Or potentially storing `Option<BidiClass>` or something)

Everything is in UTF-8 indices.